File: src/bun.js/webcore/S3Stat.zig
pub const S3Stat = struct {
    const log = bun.Output.scoped(.S3Stat, .visible);
    pub const js = jsc.Codegen.JSS3Stat;
    pub const toJS = js.toJS;
    pub const fromJS = js.fromJS;
    pub const fromJSDirect = js.fromJSDirect;

    pub const new = bun.TrivialNew(@This());

    size: u64,
    etag: bun.String,
    contentType: bun.String,
    lastModified: f64,

    pub fn constructor(globalThis: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!*@This() {
        return globalThis.throwInvalidArguments("S3Stat is not constructable", .{});
    }

    pub fn init(
        size: u64,
        etag: []const u8,
        contentType: []const u8,
        lastModified: []const u8,
        globalThis: *jsc.JSGlobalObject,
    ) bun.JSError!*@This() {
        var date_str = bun.String.init(lastModified);
        defer date_str.deref();
        const last_modified = try date_str.parseDate(globalThis);

        return S3Stat.new(.{
            .size = size,
            .etag = bun.String.cloneUTF8(etag),
            .contentType = bun.String.cloneUTF8(contentType),
            .lastModified = last_modified,
        });
    }

    pub fn getSize(this: *@This(), _: *jsc.JSGlobalObject) jsc.JSValue {
        return jsc.JSValue.jsNumber(this.size);
    }

    pub fn getEtag(this: *@This(), globalObject: *jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
        return this.etag.toJS(globalObject);
    }

    pub fn getContentType(this: *@This(), globalObject: *jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
        return this.contentType.toJS(globalObject);
    }

    pub fn getLastModified(this: *@This(), globalObject: *jsc.JSGlobalObject) jsc.JSValue {
        return jsc.JSValue.fromDateNumber(globalObject, this.lastModified);
    }

    pub fn finalize(this: *@This()) void {
        this.etag.deref();
        this.contentType.deref();
        bun.destroy(this);
    }
};

const bun = @import("bun");
const jsc = bun.jsc;

File: src/runtime/webcore/S3Stat.rs
use bun_core::String as BunString;
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult, StringJsc as _, bun_string_jsc};

bun_output::declare_scope!(S3Stat, visible);

#[bun_jsc::JsClass]
pub struct S3Stat {
    pub(crate) size: u64,
    pub(crate) etag: BunString,
    pub(crate) content_type: BunString,
    pub(crate) last_modified: f64,
}

impl S3Stat {
    pub(crate) fn constructor(global: &JSGlobalObject, _frame: &CallFrame) -> JsResult<Box<Self>> {
        Err(global.throw_illegal_constructor())
    }

    pub(crate) fn init(
        size: u64,
        etag: &[u8],
        content_type: &[u8],
        last_modified: &[u8],
        global: &JSGlobalObject,
    ) -> JsResult<Box<Self>> {
        let last_modified =
            bun_string_jsc::parse_date(&BunString::from_bytes(last_modified), global)?;

        Ok(Box::new(S3Stat {
            size,
            etag: BunString::clone_utf8(etag),
            content_type: BunString::clone_utf8(content_type),
            last_modified,
        }))
    }

    #[bun_jsc::host_fn(getter)]
    pub(crate) fn get_size(&self, _global: &JSGlobalObject) -> JSValue {
        JSValue::js_number(self.size as f64)
    }

    #[bun_jsc::host_fn(getter)]
    pub(crate) fn get_etag(&self, global: &JSGlobalObject) -> JsResult<JSValue> {
        self.etag.to_js(global)
    }

    #[bun_jsc::host_fn(getter)]
    pub(crate) fn get_content_type(&self, global: &JSGlobalObject) -> JsResult<JSValue> {
        self.content_type.to_js(global)
    }

    #[bun_jsc::host_fn(getter)]
    pub(crate) fn get_last_modified(&self, global: &JSGlobalObject) -> JSValue {
        JSValue::from_date_number(global, self.last_modified)
    }
}
